packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -953,31 +1119,36 @@ class TextPainter {
   /// To set the text style, specify a [TextStyle] when creating the [TextSpan]
   /// that you pass to the [TextPainter] constructor or to the [text] property.
   void paint(Canvas canvas, Offset offset) {
-    final double? minWidth = _lastMinWidth;
-    final double? maxWidth = _lastMaxWidth;
-    if (_paragraph == null || minWidth == null || maxWidth == null) {
+    final _TextPainterLayoutCache? layoutCache = _layoutCache;
+    if (layoutCache == null) {
       throw StateError(
         'TextPainter.paint called when text geometry was not yet calculated.\n'
         'Please call layout() before paint() to position the text before painting it.',
       );
     }
 
+    if (!layoutCache.offset.dx.isFinite || !layoutCache.offset.dy.isFinite) {
+      return;
+    }
+
     if (_rebuildParagraphForPaint) {
+      _rebuildParagraphForPaint = false;
       Size? debugSize;
       assert(() {
         debugSize = size;
         return true;
       }());
 
-      _createParagraph();
-      // Unfortunately we have to redo the layout using the same constraints,
-      // since we've created a new ui.Paragraph. But there's no extra work being
-      // done: if _needsPaint is true and _paragraph is not null, the previous
-      // `layout` call didn't invoke _layoutParagraph.
-      _layoutParagraph(minWidth, maxWidth);
+      final ui.Paragraph paragraph = layoutCache.layout._paragraph;
+      // Unfortunately even if we know that there is only paint changes, there's
+      // no API to only make those updates so the paragraph has to be recreated
+      // and redo layout.
+      layoutCache.layout._paragraph = _createParagraph(text!)..layout(ui.ParagraphConstraints(width: layoutCache.inputWidth));
+      assert(paragraph.width == layoutCache.layout._paragraph.width);
+      paragraph.dispose();
```

Comment:
Should this just happen in the paragraph setter above?

Reply:
Overall the ownership model is a bit strange. Might be a little clearer if the layoutCache would own the paragraph and dispose it when necessary...

Reply:
I think the weirdness came from ui.Paragraph, as it doesn't distinguish layout changes and paint changes, and skparagraph also suffers from this: it has to build a caching mechanism to avoid reshaping for paint-only changes.

Reply:
Ok, the model should be consistent, though. Who owns the paragraph? Who is in charge of disposing it? Right now it's inconsistent (likely due to some other suggestions I made...).

Reply:
`TextPainter` still owns the paragraph (`_TextPainterLayoutCacheWithOffset ` doesn't create paragraphs or dispose them, it's just a container) so I guess I remove the new `dispose` method.